src/debug/debug_impl.ts

```diff
@@ -0,0 +1,197 @@
+'use strict';
+
+import * as child_process from "child_process";
+import * as path from "path";
+import {
+	DebugSession,
+	InitializedEvent, TerminatedEvent, StoppedEvent, BreakpointEvent, OutputEvent, Event,
+	Thread, StackFrame, Scope, Source, Handles, Breakpoint, ThreadEvent, Variable
+} from "vscode-debugadapter";
+import { DebugProtocol } from "vscode-debugprotocol";
+
+export interface DartLaunchRequestArguments extends DebugProtocol.LaunchRequestArguments {
+	cwd: string;
+	program: string;
+	args: Array<string>;
+}
+
+export class DartDebugSession extends DebugSession {
+	private sourceFile: string;
+	private cwd: string;
+	private childProcess: child_process.ChildProcess;
+	private processExited: boolean = false;
+
+	public constructor() {
+		super();
+
+		this.setDebuggerLinesStartAt1(false);
+		this.setDebuggerColumnsStartAt1(false);
+	}
+
+	protected initializeRequest(
+		response: DebugProtocol.InitializeResponse,
+		args: DebugProtocol.InitializeRequestArguments
+	): void {
+		response.body = {
+			supportsConfigurationDoneRequest: true,
+			supportsEvaluateForHovers: true,
+			exceptionBreakpointFilters: [
+				{ filter: "All", label: "All Exceptions", default: false },
+				{ filter: "Unhandled", label: "Uncaught Exceptions", default: true }
+			]
+		};
+		this.sendResponse(response);
+	}
+
+	protected launchRequest(response: DebugProtocol.LaunchResponse, args: DartLaunchRequestArguments): void {
+		this.cwd = args.cwd;
+		this.sourceFile = path.relative(args.cwd, args.program);
+		this.sendEvent(new OutputEvent(`dart ${this.sourceFile}\n`));
+
+		this.sendResponse(response);
+
+		let debug = !args.noDebug;
+		let appArgs = [];
+		// TODO:
+		// if (debug) {
+		// 	appArgs.push("--enable-vm-service:0");
+		// 	appArgs.push("--pause_isolates_on_start=true");
+		// }
+		appArgs.push(this.sourceFile);
+		if (args.args)
+			appArgs = appArgs.concat(args.args);
+
+		// TODO: dart sdk path - send a request to the client for it
```

Comment:
We can probably fail w/ an error message that `dart` was not found, and they should adjust their path. Eventually, it would be great to be able to get the value from the user's settings. That way running and debugging will work as long as they have configured vscode w/ an sdk.